### ISSUE TITLE
Why is the value type of queryParameters in the Uri class dynamic?

### ISSUE THREAD

**Issue by myConsciousness** [dart-lang/sdk#47970]
Hi amazing developers,

The Uri class in dart:core has queryParameters as an argument, but why is the value type dynamic even though it only works with String?

When I used this function, I thought that since the type of value was dynamic, the internal processing would convert the data that was the basis of the query parameter passed as an argument into a string and generate the query parameter, but it seems that this is not the case. Or is the purpose different from the first place?

```dart
 factory Uri(
      {String? scheme,
      String? userInfo,
      String? host,
      int? port,
      String? path,
      Iterable<String>? pathSegments,
      String? query,
      Map<String, dynamic /*String|Iterable<String>*/ >? queryParameters,
      String? fragment}) = _Uri;
```

Thank you.

**Comment by julemand101** [comment]
Because it works with `String` and `Iterable<String>`. Since the type system in Dart is not capable to define such a requirement, the argument needs to be declared `dynamic`. You can see this in the comment in the code you pasted with `/*String|Iterable<String>*/` which tries to tell you this requirement as a comment.

This is also hinted in the documentation:
> When `queryParameters` is used the query is built from the provided map. Each key and value in the map is percent-encoded and joined using equal and ampersand characters. A value in the map must be either a string, or an Iterable of strings, where the latter corresponds to multiple values for the same key.

https://api.dart.dev/stable/2.15.1/dart-core/Uri/Uri.html

**Comment by myConsciousness** [comment]
Hi @julemand101 ,

Thank you, I understand the situation. 

It is true that if you want to set multiple values associated with a key, you have to express it this way in Dart language.
I was just asking because when I first used this feature, I found it a genuinely weird implementation. 

Thank you very much anyway :)

**Comment by lrhn** [comment]
There is some history in the API.
Originally it only allowed one value per key. When that was changed, the `Map` value type was changed from `String` to `dynamic` (aka: `String` or `Iterable<String>`) as a way to stay backwards compatible.
It's not great, and it became even less great with the stronger type checking of Dart 2.0.

If I had to create the API from scratch today, it wouldn't be how I'd do it. Alas, it's what we have, and breaking it today will hurt more than it helps.
